graph.hpp

```diff
@@ -0,0 +1,97 @@
+#ifndef GRAPH_H_INCLUDE
+#define GRAPH_H_INCLUDE
+#include<vector>
+#include<map>
+#include<set>
+#include<memory>
+
+typedef long long int64;
+
+class Graph {
+
+  // Nested class representing node in the graph to manage connections
+  class Node {
+    // Since this class is not used anywhere else, OK to make it a friend
+    friend class Graph;
+    // These represent incoming and outgoing edges.
+    // Reasons for having both sets:
+    // 1. When a node is deleted, this makes it finding out which edges to
+    //    delete much easier
+    // 2. This makes reverse a lot easier
+    std::shared_ptr<std::set<int64> > outgoing;
+    std::shared_ptr<std::set<int64> > incoming;
+    // A bunch of mutator methods to add/delete incoming/outgoing edges from
+    // node
+    void insertOutgoing(int64 to);
+    void insertIncoming(int64 from);
+    void eraseOutgoing(int64 to);
+    void eraseIncoming(int64 from);
+    // Two accessor methods. Return true if there exists an incoming/outgoing
+    // edge
+    bool containsEdgeFrom(int64 from);
+    bool containsEdgeTo(int64 to);
+    Node DeepCopy();
+    public:
+    Node() :
+      outgoing(new std::set<int64>()),
+      incoming(new std::set<int64>()) {}
+  };
+
+  // collection of nodes, mapping ids to nodes
+  std::map<int64, Node> nodemap;
```

Comment:
In general at Google we prefer containers to have pointers to things unless the thing itself is really tiny, like an int or other primitive type. The reason for this is that containers generally copy their elements when they run out of space, which has caused performance issues for us in the past where huge lists are being copied over. Here Node is pretty small, but it's scary to put a class in there as a non-pointer because as people add stuff to the class in the future it'll grow perniciously over time, which sucks. I think for the exercise it would be better if what you had was:

std::map<int64, std::unique_ptr<Node>> nodemap;

This will also give you some great experience with unique_ptr, which you'll use a lot at Google. unique_ptr is basically saying "I own this thing and I'll delete it when I go out of scope," which is an incredibly useful construct because if you only use unique_ptr you'll never have a memory leak.